### DISCUSSION TITLE
Can you prevent 'Any' from always matching the first overload?

### DISCUSSION THREAD

**Discussion by maflAT** [python/typing#1792] [Q&A] [answered]
Sorry for the vague and maybe misleading title. I couldn't think of a good way to succinctly describe my problem.
I'm trying to type this generic class and want to define a fall-through overload for when the specialization is Unknown/Any. 

[pyright playground](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoF0SKJAT1y1IHNseAkQCCpLm3DRuvAUPyEoYAG5oQAGzDEAJlI6y%2BgnIqLo1xDTFloWd5gGMNxevSgBhANoAVAFycEXA00T3EuABooADoYgF1Y3xYoZKgAAVV1LV0klJ00YChgMDAACno0DWB-LxQgkJKASniGqABaAD4AutCJBOiYnOT0tU1tPWYUqDyCotLyyurPWuCerma2zuWQsMiwvpiowbSM0eyJ3PzC4rKKqo9V2JaOrpWd-qj9gdZvvnoqUgcaEWWwesSgAF4PI0WOY0JZrDw0CVfv9AVFZo0GiwAMRQbyI5QFABEKLIaIxDSJ2DcRJBbz2VPBnQcBHQDiIJROWR0UAATFifgxUUD7nSJLtemDIe5ocxYfCbMihWS0Ojrg0Bbj8bg0ISoCTlQDVRSqVgaWKIu9Yozmay0OyoCVgJYNJwABbgBD8N1QBwuNACliko2LPYQqEC%2BVWRXB8nqzV4glgYmx43q03mwKvXo2qAAd3A8k5I25UAAjAmAMpusAIDQ8iDUBw%2BpBu3XOjQaVqtz3e33%2ByIAI2QJA0eeIXDcjaQzbQblb7awID%2ByhLY0OzCAA)

```python
from __future__ import annotations

from typing import Any
from typing import overload
from typing import reveal_type


class C[T: tuple[Any, ...]]:
    @overload
    def foo(self: C[tuple[()]]) -> tuple[Any]: ...
    @overload
    def foo(self: C[tuple[Any]]) -> tuple[Any, Any]: ...
    @overload
    def foo(self: C[Any]) -> tuple[Any, ...]: ...


instance: C[tuple[Any]] = C()
reveal_type(instance.foo())
# Type of "instance.foo()" is "tuple[Any, Any]" => correct (overload 2)

instance: C[tuple[Any, Any]] = C()
reveal_type(instance.foo())
# Type of "instance.foo()" is "tuple[Any, ...]" => correct (fall through case)

instance: C[Any] = C()
reveal_type(instance.foo())
# Type of "instance.foo()" is "tuple[Any]" => wrong (overload 1)
# Should match the fall-through case, but always matches the first overload.
``` 

The perfect solution would be 
```python
def foo(self) -> tuple[Any, *T]: ...
``` 
but that is not allowed in this context.

Is there any way i can correctly type hint this class, so that an unknown specialization will match the fall-through case?

**Comment by erictraut** [comment, marked as answer]
Overload resolution behavior isn't currently specified in the typing specification, so you will may see differences between Python type checkers — especially when `Any` is involved. In your code sample, the behaviors of pyright and mypy appear to match. For a description of pyright's overload resolution behavior, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads).

I don't understand the full constraints of your use case, but a TypeVarTuple might be preferable to a TypeVar. Here's how that might look.

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChRIokBPXLUgc2zwJEAgqXYAaKGABuaEABswxACaNGAYznF69KAGEA2gCoAKvQC6ALkZQbUAALTZC5ddtK0wKMDBgAFPTQ5YAs9fV8ASjMoAB9QkXYzcKgAWgA%2BNgRcOTQjU0tXW1sAOhLVQocZeUUVQvdPbz8AoKS0jKyc%2BIkSovzCwu7VbnoqUjU0AEYoAF440TMIxnQZYjkYDlw0XyGRsfGihoikgGI27P14s0GGHbQAJmnZ8SgLhaW0FbXOTe2yMdv9nyHKAnFDtc6iLolS6MH6jNAAZgeuleaGWq3W32uvwRAL84WOpxyAFUANYUADupEhPUYQA)

```python
from __future__ import annotations
from typing import Any, overload

class C[*Ts]:
    @overload
    def foo(self: C[()] | C[Any]) -> tuple[*Ts]:
        ...

    @overload
    def foo(self) -> tuple[Any, ...]:
        ...

instance1 = C[Any]()
reveal_type(instance1.foo()) # tuple[Any]

instance2 = C[Any, Any]()
reveal_type(instance2.foo()) # tuple[Any, ...]

instance3 = C()
reveal_type(instance3.foo()) # tuple[Uknown, ...]
```

If you simply want to prepend an `Any` element to the tuple, you can do the following:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChRIokBPXLUgc2zwJEAgqXaNGAYwA2xevSgBhANoAqACr0AugC5GUfVAAmaYFGBgwACnpopwAJRQAtAD42CXFLRKR7ADRQKiie3upaOnoGUQB0seLc9FSkEmgAjFAAvIqqwV5KlvaamgWM6ABuaMRSMBy4aJYJSSmp0eZW9o5QAMTuIT6imvEMTWgATJnZvsX2pWgVVTWc9Y1kKaOtFgWdPbnevgFTQ4mraADME8r7UFMl5ZXVtcvDJ6cb7du9eVffA0cjACwXW5ze6LOoNZ7JND-N5bfQ7DxfUQBIKI7wAVQA1hQAO6kAKxaJFIA)

```python
from __future__ import annotations
from typing import Any

class C[*Ts]:
    def foo(self) -> tuple[Any, *tuple[*Ts]]:
        ...

instance1 = C[*tuple[()]]()
reveal_type(instance1.foo())  # tuple[Any]

instance2 = C[Any]()
reveal_type(instance2.foo())  # tuple[Any, Any]

instance3 = C[Any, Any]()
reveal_type(instance3.foo())  # tuple[Any, Any, Any]

instance4 = C()
reveal_type(instance4.foo())  # tuple[Any, *tuple[Uknown, ...]]
```

**Reply by maflAT** [reply]
Hello Eric, thank you for answering. 
Yes, I originally tried your solution with a TypeVarTuple. 
Unfortunately, that does not allow bounds on the types. This is why I tried the more verbose solution with multiple overloads.

**Reply by maflAT** [reply]
Also, I had no idea you could instantiate a class from a generic alias like this `C[Any]()`. 😮 
Is this a new feature?

**Reply by erictraut** [reply]
> Is this a new feature?

No, I think this has been around since PEP 484 was implemented.
